docs: transfer of text about runtime uploads (#30)
seedcase-project/seedcase-sprout: docs/design/runtime-view.qmd

---
title: "Runtime view"
---

This section describes the concrete behaviour, interactions, and
pathways that data take within Seedcase. "Runtime" in this case refers
to how the software works "in action".

## Login and Authentication

Almost all users will need to log into the Seedcase-managed Data
Resources. The steps for logging in and having their permission levels
checked follows the sequence described in the figure below.

![Login and authentication sequence of a registered user.](images/runtime-login-sequence.png)

## Data Input

The overall aim of this section is to describe the general path that
data takes through a Seedcase Data Resource, from input into the final
output. Specifically, these items are described as:

-   *Input*: Because we currently focus on health research, the type of
    input data and metadata will be what is typically generated from
    health studies.
-   *Output*: The final output object would be the input data stored
    together as a single database, or at least multiple databases and
    files explicitly linked in such a way that it conceptually

>     represents a single database.

Author: This was more back when we weren't completely sure of the db structure, I think we can cull that part of the sentence.

-   *Path*: The computational and programmatic steps the input data and
    metadata takes from being uploaded by a human (and potentially
    automatically by a program) into Seedcase, passing through quality
    control checks, adding to the changelog, and storage into the
    database.

### Expected Type of Input Data

Given the (current) focus on health data as well as our own experiences,
we make some assumptions about the type of data that will be input into

> Seedcase. Health data tends to consist of specific types of data:

Author: That is also my understanding

> <!---

Author: I wasn't sure if it was relevant for Sprout, or if we should take it out.

> <!---

Reviewer: We can keep it in to describe our expected data sources.

-   **Clinical**: This data is typically collected during patient visits
    to doctors. Depending on the country or administrative region, there
    will likely already be well-established data processing and storage
    pipelines in place. --->
-   **Register**: This type of data is highly dependent on the country
    or region. Generally, this data is collected for national or
    regional administrative purposes, such as, recording employment
    status, income, address, medication purchases, and diagnoses. Like

>     the routine clinical data, the pipelines in place for processing and

Author: I'm partial to keeping the section that I commented out, just worried about the length of text.

    generated from sample analytic techniques often produce large
    volumes of data per person. Samples may be generated in larger
    established laboratories or in smaller research groups, depending on
    how what analytic technology is used and how new it is. The
    structure and format of the generated data also tends to be highly
    variable and depends heavily on the technology used, sometimes
    requiring specialized software to process and output. --->
-   **Survey or questionnaire**: This type of data is often done based
    on a given study's aims and research questions. There are hundreds
    of different questionnaires that can have highly specific purposes
    and uses for their data. They are also highly variable in the volume
    of data collected based on the survey, and on the format of the
    data.

### Expected Flow of Input Data

The above described data tends to fit into, mostly, two categories for
data input.

-   *Routine or continuous collection*, where ingested data into
    Seedcase would occur as soon as the data was collected from one
    "observational unit"[^1] or very shortly afterwards. Clinical data

>     as well as survey or questionnaire data may likely fall under this

Reviewer: We'll keep the clinical data section.

    category.
-   *Batch collection*, where ingested data occurs some time after the
    data was collected and from multiple observational units. Biological
    sample data would fall under this category, since laboratories
    usually run several samples at once and input data after internal

>     quality control checks and machine-specific data processing. While

Reviewer: Keeping biological sample as well.

    register-based data does get collected continuously, direct access
    to it is only given on a batch basis, usually once every year.
    Survey data may also come in batches, depending on the questionnaire
    and software used for its collection.

[^1]: Observational unit is the "entity" that the data was collected
    from at a given point in time, such as a human participant in a
    cohort study or a rat in an animal study at a specific time point.

For sources of data from routine collection with well-established data
input processes, the data input pipeline would likely involve
redirecting these data sources from their generation into Seedcase via a
direct call to the API so the data continues on to the backend and
eventual data storage.

Sources of data that don't have well-established data input processes,
such as from hospitals or medical laboratories, would need to use the

> Seedcase data batch-input Web Portal. This Portal would only accept data
> that is in a pre-defined format (as determined and created by the data
> owner) and would include documentation, and potentially automation
> scripts, on how to pre-process the data prior to uploading it.

Reviewer 2: ```suggestion
Sprout data batch-input Web Portal. This Portal only accepts data
that is in a pre-defined format (as determined and created by the Data Management 
Administrators) that includes documentation, and potentially automation
scripts on how to pre-process the data prior to uploading it.
```
For my understanding: 
- What does the documentation mentioned here include? Is that the *definitions* we included in the schemas described in the user flow diagrams?
- Will the "automation scripts on how to pre-process the data prior to uploading it" be something that users Data Management Administrators will be able to add into Sprout somehow? This is currently not mentioned in the user flow diagrams

Author: I'm no longer sure what I meant when I wrote this, may just take out everything after the set of brackets.

These uploaded files might be a variety of file types, like `.csv`,
`.xls`, or `.txt`). Only users with the correct permission levels are

> allowed to upload data. In either case it will be the administrator who
> will be doing the initial upload, as that will entail setting up tables
> and allocating space in the raw data file storage. The second way of

Reviewer 2: ```suggestion
allowed to upload data. It will be the Data Access Administrator who
will be doing the initial upload, as that will entail setting up table schemas
and allocating space in the raw data file storage. The second way of
```

getting data into the Data Resource is by manually enter it by an

> authorized user.

Reviewer 2: ```suggestion
authorized Data Contributor.
```

Reviewer 2: Maybe even add link to the description of the user roles?

Author: Good idea.

> Once the data is submitted through the Portal, it would get sent in an

Reviewer 2: ```suggestion
Once the data is submitted through the Portal, it is sent in an
```

> encrypted, legally-compliant format to the server and stored in the way

Reviewer 2: ```suggestion
encrypted, legally-compliant format to a server and stored in the way
```

> defined by the API and common data model.

Reviewer 2: Do we need a definition or description of what a "common data model" is here?

Author: I'm leaning towards a model where we skip the common data model for now, we can add them in version 15 or so...

Reviewer: See #50

> ### Preparation for Getting Data into Sprout

Reviewer 2: ```suggestion
### Upload Data to Sprout
```

> An approved user will open the login screen in the Web Portal. They

Reviewer 2: ```suggestion
An approved user, i.e., a Data Access Administrator or a Data Contributor, will open the login screen in the Web Portal. They
```

will enter their credentials which will be transmitted to the API layer.
The API Security layer will check with the list of users and permissions
in the database and confirm that the specific user has permission to
enter data into a specific table (or set of tables) in the database.

Once this check is complete the frontend will receive permission from

> the API Security layer to display the data entry/upload options.

Reviewer 2: ```suggestion
the API Security layer to display the data entry/upload options for this kind of user role.
```

> Before any of the actions described below can be done, it is expected that appropriate table schemas or entry forms have been created by one or more administrators of the system.  This process is described elsewhere.

Reviewer 2: ```suggestion
Before any of the actions described below can be done, it is expected that appropriate table schemas or entry forms have been created by one or more administrators of the system. This process is described elsewhere.
```

Reviewer 2: Could it make sense to go through the preparation for upload steps (i.e., table schema creation) here?

Author: I think so.

<!--TODO change elsewhere above to the actual location of where we describe table schema and data entry form creation-->

> #### Batch Upload of Data 

Reviewer 2: This and maybe the section above could maybe use some additional descriptions using the user flow diagrams? :)

Author: I admit that I'm losing the overview of this text, is there any way that we can commit the changes done so far, and then revisit the comments left afterwards?  Maybe we can look at that either Wednesday or Thursday if we are both in the office?

Reviewer 2: Yeah, me too. Let's look at it together!

Reviewer: I'll add a TODO item here.

> The user has selected a valid table schema to use, and have uploaded the file to the holding area. This prompts the system to check that the data in the file match the schema in the database on headers and data type. If this validation is successful then the system will inform the user about how many rows of data it found and validated. If the user is in agreement, then the system will write the data into the relevant table and display a confirmation back to the user. Should the user disagree with the number of rows then they cancel the upload and take the file away to investigate why the system can't see the correct number of rows, this is an action which happens outside of Seedcase.

Reviewer 2: What is meant by a "valid table schema" here? Is valid = existing in this context?

Author: I'm thinking a table schema which has been created and is ready to use.  Probably a better word than valid for it.

> ![Logged in user who choses to use the batch upload function with existing table schema.](/design/images/user-flow-5-data-upload.png){#fig-batch-data-entry}

Reviewer 2: ```suggestion
![Logged In User Who Chooses to Use the Batch Upload Function with Existing Table Schema.](/design/images/user-flow-data-upload.png){#fig-batch-data-entry}
```

Reviewer 2: We seem to have a bit of an overlap between this post and the user-flow post with this post using one of the user flow diagrams? Should we separate them completely by having this post be at a more technical level (i.e., with sequence diagrams only)?

<!--TODO Ensure that the link above will still work once SKB has finished updating the diagrams-->

#### Manual Data Entry: Done in One Session

The user completes all fields in the form and clicks "Save and Submit". This sends
the data to the API layer where it is confirmed as valid, parcelled up
and submitted to the database. The database will then write the data
into a new record in the table (or tables). Once done the database will
confirm successful entry of data to the API which will in turn send the
confirmation back to the user via the frontend.

> ![Logged in user who manually writes a new row to the Data
> Resource.](/design/images/runtime-manual-data-entry.png){#fig-manual-data-entry}

Reviewer 2: ```suggestion
![Logged In User Who Manually Writes a New Row to the Data
Resource.](/design/images/runtime-manual-data-entry.png){#fig-manual-data-entry}
```

<!--TODO convert puml file to png so that the link above works-->

#### Manual Data Entry: Done in Multiple Sessions

There may be situations where an approved user will be prevented from
completing the data entry form in one session. In that case it would be
beneficial if there is an option of saving the data as it is, and be
able to return to the data entry at a later time. Much of the initial
workflow is the same as above, until the user is interrupted and selects
"Save" instead of "Save and Submit". This will send the data to the API with
a flag showing that fields may be incomplete, thus preventing the API
from rejecting the data due to NULL values. The API will submit the data
to the database along the incomplete flag.

> When the user at a later time goes back to the data entry they will be

Reviewer 2: ```suggestion
When the Data Contributor goes back to the data entry at a later time, they will be
```

presented with the option of completing any incomplete records as well
as entering new data. If they click on "Complete Records" they are shown
the records that they have started but not submitted. Once they select a
partially completed record the frontend will request the currently
completed items from the database via the API layer before displaying
the entry form with the completed fields.

Once the user has completed more data they can either click on "Save" or
"Save and Submit". The first option will put them back to the top of this
workflow, the second will send the data back to the API layer for
validation. Once the data is validated it will be submitted to the
database. The database will then write the data into a new record in the
table (or tables) and update the flag to show the record is complete.
Once done the database will confirm successful entry of data to the API
which will in turn send the confirmation back to the user via the front
end.

> ![Logged in user enters data manually in more than one
> session](/design/images/runtime-manual-data-update.png){#fig-manual-data-update}

Reviewer 2: ```suggestion
![Logged In User Enters Data Manually in More Than One
Session](/design/images/runtime-manual-data-update.png){#fig-manual-data-update}
```
